packages/components/src/input-control/state.js

```diff
@@ -0,0 +1,254 @@
+ *
+ * @param  {...any} fns State reducers.
+ * @return {Function} The single combined stateReducer.
+ */
+export const combineStateReducers = ( ...fns ) => {
+	const reducers = fns.flat( Infinity );
+
+	return ( ...args ) => {
+		return reducers.reduceRight( ( state, fn ) => {
+			return { ...state, ...fn( ...args ) };
+		}, {} );
+	};
+};
+
+const memoizedCombineStateReducers = memize( combineStateReducers );
+
+/**
+ * Creates a reducer that opens the channel for external state subscription
+ * and modification.
+ *
+ * This technique uses the "stateReducer" design pattern:
+ * https://kentcdodds.com/blog/the-state-reducer-pattern/
+ *
+ * @param {Function} combinedStateReducers A custom reducer that can subscribe and modify state.
+ * @return {Function} The reducer.
+ */
+function inputControlStateReducer( combinedStateReducers ) {
+	return ( state, action ) => {
+		const nextState = { ...state };
+		const { type, payload } = action;
+
+		switch ( type ) {
+			/**
+			 * Keyboard events
+			 */
+			case actionTypes.PRESS_UP:
+				nextState.isDirty = false;
+				break;
+
+			case actionTypes.PRESS_DOWN:
+				nextState.isDirty = false;
+				break;
+
+			/**
+			 * Drag events
+			 */
+			case actionTypes.DRAG_START:
+				nextState.isDragging = true;
+				break;
+
+			case actionTypes.DRAG_END:
+				nextState.isDragging = false;
+				break;
+
+			/**
+			 * Input events
+			 */
+			case actionTypes.CHANGE:
+				nextState.error = null;
+				nextState.value = payload.value;
+
+				if ( state.isPressEnterToChange ) {
+					nextState.isDirty = true;
+				}
+
+				break;
+
+			case actionTypes.SUBMIT:
+				nextState.value = payload.value;
+				nextState.isDirty = false;
+				break;
+
+			case actionTypes.RESET:
+				nextState.error = null;
+				nextState.isDirty = false;
+				nextState.value = payload.value || state.initialValue;
+				break;
+
+			case actionTypes.UPDATE:
+				if ( payload.value !== state.value ) {
+					nextState.value = payload.value;
+					nextState.isDirty = false;
+				}
+				break;
+
+			/**
+			 * Validation
+			 */
+			case actionTypes.INVALIDATE:
+				nextState.error = payload.error;
+				break;
+		}
+
+		if ( payload.event ) {
+			nextState._event = payload.event;
+		}
+
+		/**
+		 * Send the nextState + action to the combinedReducers via
+		 * this "bridge" mechanism. This allows external stateReducers
+		 * to hook into actions, and modify state if needed.
+		 */
+		return combinedStateReducers( nextState, action );
+	};
+}
+
+/**
+ * A custom hook that connects and external stateReducer with an internal
+ * reducer. This hook manages the internal state of InputControl.
+ * However, by connecting an external stateReducer function, other
+ * components can react to actions as well as modify state before it is
+ * applied.
+ *
+ * This technique uses the "stateReducer" design pattern:
+ * https://kentcdodds.com/blog/the-state-reducer-pattern/
+ *
+ * @param {Function} stateReducer An external state reducer.
```

Comment:
This type declaration is incorrect, since the hook expects an array of functions.

Now, I do think that `Function` is the right way to go. In contrast, I think all components expecting/passing state reducers should be very strict about only passing one function.

This means that, whenever, a component such as InputControl or NumberControl is merging its own reducer rules with its parent-provided rules, it should not be concatenating as `[ numberControlStateReducer, stateReducer ]` and should instead be passing a combined function. So we'd have a chain like:

```js
  <UnitControl stateReducer={ myAppStateReducer } />
= <NumberControl stateReducer={ combineStateReducers( numberControlStateReducer, props.stateReducer ) } />
= <InputControl stateReducer={ combineStateReducers( inputControlStateReducer, props.stateReducer ) } />
```

Reply:
> In contrast, I think all components expecting/passing state reducers should be very strict about only passing one function.

Ah, interesting! (Also, thank you for the type def spot!)

My thinking was for the `combineStateReducers` to automatically handle the merging of multiple functions. Making it slightly less verbose for components - especially for nested cases. 

Perhaps keeping it more explicit is better to start?

If we start using this pattern/these functions elsewhere, we can consolidate and refactor then.

Reply:
So my reasoning here follows what is standard practice in Redux: data domains of an application specify their reducers as plain functions of type `State -> Action -> State`, and some "controller" logic of the application combines these in succession.

An example:

```js
// app.js
const reducer = combineReducers( {
  users,
  posts,
  ui,
} );

// ui.js
const reducer = combineReducers( {
  sidebars,
  notices,
} );

// etc
```

That way, each level is responsible for the data model at its scope. So `appReducer` doesn't care how `uiReducer` is built — it just needs it to be a function of type `State -> Action -> State`.

Understanding that StateReducer isn't the same pattern, it does have many parallels. Since StateReducer is based on inversion of control, then it would follow that the innermost (InputControl) doesn't care how NumberControl packaged its reducer, nor how the outermost UnitControl did.

We don't need to reach consensus now if it helps keep things moving, but this does seem like a pertinent point to get right.